pkg/gateway/model/route_match.go

```diff
@@ -0,0 +1,115 @@
+package model
+
+import (
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/skygeario/skygear-server/pkg/core/config"
+)
+
+const maxRouteAttempt = 5
+
+type RouteMatch struct {
+	Route config.DeploymentRoute
+	Path  string
+}
+
+func matchRoutePath(reqPath string, routes []config.DeploymentRoute) *RouteMatch {
+	// convert path to /path/ format
+	testReqPath := appendtrailingSlash(reqPath)
+
+	var matchedRoute *config.DeploymentRoute
+	matchedLen := 0
+	for _, r := range routes {
+		if reqPath == r.Path {
+			matchedRoute = &r
+			break
+		}
+		testPath := appendtrailingSlash(r.Path)
+		if !strings.HasPrefix(testReqPath, testPath) {
+			continue
+		}
+		if len(r.Path) > matchedLen {
+			matchedLen = len(r.Path)
+			// We must copy the route because r is
+			// reused in all iterations.
+			copied := r
+			matchedRoute = &copied
+		}
+	}
+
+	if matchedRoute == nil {
+		return nil
+	}
+
+	routePath := strings.TrimSuffix(matchedRoute.Path, "/")
+	matchPath := strings.TrimPrefix(reqPath, routePath)
+	if len(matchPath) == 0 {
+		matchPath = "/"
+	}
+	return &RouteMatch{
+		Route: *matchedRoute,
+		Path:  matchPath,
+	}
+}
+
+func MatchRoute(reqPath string, routes []config.DeploymentRoute) *RouteMatch {
+	attempt := 0
+	for attempt < maxRouteAttempt {
+		attempt++
+
+		match := matchRoutePath(reqPath, routes)
+		if match == nil {
+			return nil
+		}
+		if match.Route.Type == DeploymentRouteTypeStatic {
+			config := RouteTypeConfig(match.Route.TypeConfig)
+			pathMapping := config.AssetPathMapping()
+			assetPath := path.Clean("/" + reqPath)
```

Comment:
Should it be `match.Path` instead of `reqPath` here?

Reply:
Or the `asset_path_mapping` keys will have the path prefix?

Reply:
Fixed. Should be using `match.Path`.